### PR TITLE
[SPARK-44748][SQL] Query execution for the PARTITION BY clause in UDTF TABLE arguments

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

This PR implements query execution support for for the PARTITION BY and ORDER BY clauses for UDTF TABLE arguments.

* The query planning support was added in [1] and [2] and [3]. After those changes, the planner added a projection to compute the PARTITION BY expressions, plus a repartition operator, plus a sort operator.
* In this PR, the Python executor receives the indexes of these expressions within the input table's rows, and compares the values of the projected partitioning expressions between consecutive rows.
* When the values change, this marks the boundary between partitions, and so we call the UDTF instance's `terminate` method, then destroy it and create a new one for the next partition.

[1] https://github.com/apache/spark/pull/42100
[2] https://github.com/apache/spark/pull/42174
[3] https://github.com/apache/spark/pull/42351

Example:

```
# Make a test UDTF to yield an output row with the same value
# consumed from the last input row in the input table or partition.
class TestUDTF:
    def eval(self, row: Row):
	self._last = row['input']
	self._partition_col = row['partition_col']

    def terminate(self):
	yield self._partition_col, self._last

func = udtf(TestUDTF, returnType='partition_col: int, last: int')
self.spark.udtf.register('test_udtf', func)
self.spark.sql('''
    WITH t AS (
        SELECT id AS partition_col, 1 AS input FROM range(0, 2) UNION ALL
        SELECT id AS partition_col, 2 AS input FROM range(0, 2)
    )
    SELECT *
    FROM test_udtf(TABLE(t) PARTITION BY partition_col ORDER BY input)
    ''').collect()

> [Row(partition_col=0, last=2), (partition_col=1, last=2)]
```

### Why are the changes needed?

This brings full end-to-end execution for the PARTITION BY and/or ORDER BY clauses for UDTF TABLE arguments.

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR adds end-to-end testing in `test_udtf.py`.

### CODE REVIEW

**Comment by dtenedor** [comment]
CC @ueshin @allisonwang-db @HyukjinKwon

File: python/pyspark/worker.py [hunk 1]
```diff
@@ -564,23 +566,66 @@ def read_udtf(pickleSer, infile, eval_type):
...
+
     # Instantiate the UDTF class.
-    try:
-        udtf = handler()
-    except Exception as e:
-        raise PySparkRuntimeError(
-            error_class="UDTF_EXEC_ERROR",
-            message_parameters={"method_name": "__init__", "error": str(e)},
-        )
+    def create_udtf_class_instance(return_type):
```
Comment (ueshin) [inline python/pyspark/worker.py:589]: Seems like `return_type` is not be used in this function?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -2117,11 +2120,25 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
                   tableArgs.size)
               }
               val alias = SubqueryAlias.generateSubqueryName(s"_${tableArgs.size}")
+              // Propagate the column indexes for TABLE arguments to the PythonUDTF instance.
+              val tvfWithTableColumnIndexes: LogicalPlan = tvf match {
+                case g @ Generate(p: PythonUDTF, _, _, _, _, _) =>
+                  functionTableSubqueryArgs.headOption.map { tableArg =>
+                    val indexes = PythonUDTFPartitionColumnIndexes(
+                      numPartitionChildIndexes = tableArg.partitioningExpressionIndexes.length,
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2128]: Do we need this? It seems pretty easy to get by `partitionChildIndexes.length`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -2117,11 +2120,25 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
                   tableArgs.size)
               }
               val alias = SubqueryAlias.generateSubqueryName(s"_${tableArgs.size}")
+              // Propagate the column indexes for TABLE arguments to the PythonUDTF instance.
+              val tvfWithTableColumnIndexes: LogicalPlan = tvf match {
+                case g @ Generate(p: PythonUDTF, _, _, _, _, _) =>
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2125]: We also need to take care of `UnresolvedPolymorphicPythonUDTF`?

File: python/pyspark/worker.py [hunk 4]
```diff
@@ -564,23 +566,66 @@ def read_udtf(pickleSer, infile, eval_type):
...
+    # partitioning semantics.
+    def check_partition_boundaries(arguments):
+        if num_partition_child_indexes == 0 or udtf_state.prev_arguments is None:
+            udtf_state.prev_arguments = arguments
+            return False
+        cur_table_arg = [arg for arg in arguments if type(arg) is Row][0]
+        prev_table_arg = [arg for arg in udtf_state.prev_arguments if type(arg) is Row][0]
+        cur_partition_values = [cur_table_arg[i] for i in partition_child_indexes]
+        prev_partition_values = [prev_table_arg[i] for i in partition_child_indexes]
+        udtf_state.prev_arguments = arguments
```
Comment (ueshin) [inline python/pyspark/worker.py:623]: This should be set before `eval` is called?

https://github.com/apache/spark/blob/58783339e51e7c6198586b32ecd4af3a971b7097/python/pyspark/worker.py#L689
https://github.com/apache/spark/blob/58783339e51e7c6198586b32ecd4af3a971b7097/python/pyspark/worker.py#L767

File: python/pyspark/worker.py [hunk 5]
```diff
@@ -564,23 +566,66 @@ def read_udtf(pickleSer, infile, eval_type):
...
             "the query again."
         )
 
+    # Inspects the values of the projected PARTITION BY expressions, if any. Returns true when
+    # these values change, in which case the caller should invoke the 'terminate' method on the
+    # UDTF class instance and then destroy it and creates a new one to implement the desired
+    # partitioning semantics.
+    def check_partition_boundaries(arguments):
+        if num_partition_child_indexes == 0 or udtf_state.prev_arguments is None:
+            udtf_state.prev_arguments = arguments
```
Comment (ueshin) [inline python/pyspark/worker.py:617]: ditto.

File: python/pyspark/worker.py [hunk 6]
```diff
@@ -564,23 +566,66 @@ def read_udtf(pickleSer, infile, eval_type):
...
+            return False
+        cur_table_arg = [arg for arg in arguments if type(arg) is Row][0]
+        prev_table_arg = [arg for arg in udtf_state.prev_arguments if type(arg) is Row][0]
+        cur_partition_values = [cur_table_arg[i] for i in partition_child_indexes]
+        prev_partition_values = [prev_table_arg[i] for i in partition_child_indexes]
+        udtf_state.prev_arguments = arguments
+        change_partitions = any(
+            (1 for k, v in zip(cur_partition_values, prev_partition_values) if k != v)
+        )
+        return change_partitions
```
Comment (ueshin) [inline python/pyspark/worker.py:627]: nit: 
```suggestion
        return any(k != v for k, v in zip(cur_partition_values, prev_partition_values))
```

**Review by dtenedor — COMMENTED** [review]
Thanks @ueshin for your helpful comments!

File: python/pyspark/worker.py (hunk 4 above)
Comment (ueshin) [reply python/pyspark/worker.py:623]: When `changed_partitions` is `True`, `create_udtf_class_instance` is called to renew the udtf instance, and also `udtf_state.prev_arguments` is set to `None` in it. So `check_partition_boundaries` will always return `False` for the next iteration?

File: python/pyspark/worker.py [hunk 7]
```diff
@@ -621,22 +663,30 @@ def verify_result(result):
...
                     # iterator is a tuple in the form of (pandas.DataFrame, arrow_return_type).
-                    yield from eval(*[a[o] for o in arg_offsets])
+                    arguments = [a[o] for o in arg_offsets]
+                    changed_partitions = check_partition_boundaries(arguments)
+                    if changed_partitions:
+                        # Call 'terminate' on the UDTF class instance, if applicable.
+                        # Then destroy the UDTF class instance and create a new one.
+                        if udtf_state.terminate is not None:
+                            yield from udtf_state.terminate()
+                        create_udtf_classs_instance(return_type)
```
Comment (ueshin) [inline python/pyspark/worker.py:681]: Need to update here, too.
```suggestion
                        create_udtf_class_instance()
```

File: python/pyspark/worker.py (hunk 4 above)
Comment (dtenedor) [reply python/pyspark/worker.py:623]: Reviewing the code again, you are right. We should be manually setting `udtf_state.prev_arguments = None` at the start just once, and then not doing that in `create_udtf_class_instance`. I made this change.

I tried to think of a unit test that would have caught this bug, but `test_udtf_with_table_argument_and_partition_by_and_order_by` should have already caught it; I am not sure why it didn't before. Leaving that test as-is for now.

**Review by dtenedor — COMMENTED** [review]
Thanks @ueshin for your reviews!!

File: python/pyspark/worker.py [hunk 8]
```diff
@@ -582,6 +583,56 @@ def read_udtf(pickleSer, infile, eval_type):
...
+    create_udtf: function
+        Function to create a new instance of the UDTF to be invoked.
+    partition_child_indexes: list
+        List of integers identifying zero-based indexes of the columns of the input table that
+        contain projected partitioning expressions. This class will inspect these values for each
+        pair of consecutive input rows. When they change, this indicates the boundary between two
+        partitions, and we will invoke the 'terminate' method on the UDTF class instance and then
+        destroy it and create a new one to implement the desired partitioning semantics.
+    """
+    class UDTFWithPartitions:
```
Comment (ueshin) [inline python/pyspark/worker.py:601]: nit: In Python, a class comment is placed after `class Xxxx:`:

```py
class UDTFWithPartitions:
    """
    This implements ...
    """

    def __init__( ... ):
...
```

File: python/pyspark/worker.py [hunk 9]
```diff
@@ -582,6 +583,56 @@ def read_udtf(pickleSer, infile, eval_type):
...
+        def __init__(self, create_udtf, partition_child_indexes):
+            self._create_udtf = create_udtf
+            self._udtf = create_udtf()
+            self._prev_arguments = None
+            self._partition_child_indexes = partition_child_indexes
+
+        def eval(self, *args, **kwargs):
+            changed_partitions = _check_partition_boundaries(args + list(kwargs.values()))
```
Comment (ueshin) [inline python/pyspark/worker.py:609]: `self._check_partition_boundaries( ... )`?

File: python/pyspark/worker.py [hunk 10]
```diff
@@ -582,6 +583,56 @@ def read_udtf(pickleSer, infile, eval_type):
...
+            if changed_partitions:
+                yield self._udtf.terminate()
```
Comment (ueshin) [inline python/pyspark/worker.py:611]: Need to check whether `self._udtf` has `terminate`?

File: python/pyspark/worker.py (hunk 8 above)
Comment (dtenedor) [reply python/pyspark/worker.py:577]: Sounds good, done.

File: python/pyspark/worker.py (hunk 9 above)
Comment (dtenedor) [reply python/pyspark/worker.py]: Done.

**Comment by dtenedor** [comment]
@ueshin @allisonwang-db I rewrote this to implement the partition boundary checks using a wrapper UDTF per Allison's suggestion.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 11]
```diff
@@ -2118,11 +2121,32 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
...
+                  }.getOrElse {
+                    g
+                  }
+                case g @ Generate(p: UnresolvedPolymorphicPythonUDTF, _, _, _, _, _) =>
+                  functionTableSubqueryArgs.headOption.map { tableArg =>
+                    val indexes = PythonUDTFPartitionColumnIndexes(
+                      tableArg.partitioningExpressionIndexes)
+                    g.copy(generator = p.copy(pythonUDTFPartitionColumnIndexes = Some(indexes)))
+                  }.getOrElse {
+                    g
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:2140]: We can combine this with the one above:
```
p : (PythonUDTF | UnresolvedPolymorphicPythonUDTF)
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchEvalPythonUDTFExec.scala [hunk 12]
```diff
@@ -134,6 +134,13 @@ object PythonUDTFRunner {
             dataOut.writeBoolean(false)
         }
     }
+    udtf.pythonUDTFPartitionColumnIndexes match {
+      case Some(partitionColumnIndexes) =>
+        dataOut.writeInt(partitionColumnIndexes.partitionChildIndexes.length)
+        partitionColumnIndexes.partitionChildIndexes.foreach(dataOut.writeInt)
```
Comment (allisonwang-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchEvalPythonUDTFExec.scala:140]: can we assert partitionColumnIndexes.partitionChildIndexes.length > 0 here?

**Review by dtenedor — COMMENTED** [review]
Thanks @allisonwang-db for your review, responded to your comments!

**Review by ueshin — APPROVED** [review]

**Comment by ueshin** [comment]
The failed tests seem not related to this.

**Comment by ueshin** [comment]
Thanks! merging to master.

**Pull request closed without merging**